doubly_linked_list/doubly_linked_list.py

```diff
@@ -38,27 +38,55 @@ class DoublyLinkedList:
   def __init__(self, node=None):
     self.head = node
     self.tail = node
+    self.max = 0
 
   def add_to_head(self, value):
-    pass
+    if value > self.max:
+      self.max = value
+    new_link = ListNode(value)
+    if self.head == None:
+      self.head = new_link
+    else:
+      self.head.insert_before(value)
+      self.head = self.head.prev
 
   def remove_from_head(self):
+    if self.head == None:
```

Comment:
what happens if you happen to be removing the max value here?

Reply:
good point!

Reply:
I think with this implementation get_max is O(1) but removing nodes would have to become O(n). Since to reliably get the second biggest max value after removing the max node you'd have to search the entire list.
Well it'd only become O(n) if you remove the node with the max value...

Reply:
on the right track! so removing a node would _always_ be O(1), but then _finding_ the max value would be O(n) (because, you're right, you'd have to search the entire list.)
my concern is also that you make sure you set `self.max` to it's new value when you do this deletion :)